vortex-array/benches/iter.rs

```diff
@@ -0,0 +1,100 @@
+use criterion::{criterion_group, criterion_main, BatchSize, Criterion};
```

Comment:
These are the results on my laptop:

```
std_iter_no_option      time:   [46.625 µs 47.232 µs 47.982 µs]
Found 9 outliers among 100 measurements (9.00%)
  5 (5.00%) high mild
  4 (4.00%) high severe

std_iter                time:   [394.24 µs 398.92 µs 403.92 µs]
Found 11 outliers among 100 measurements (11.00%)
  2 (2.00%) low severe
  1 (1.00%) low mild
  2 (2.00%) high mild
  6 (6.00%) high severe

vortex_iter             time:   [746.95 µs 750.73 µs 754.90 µs]

vortex_iter_flat        time:   [2.6330 ms 2.6477 ms 2.6635 ms]
Found 8 outliers among 100 measurements (8.00%)
  7 (7.00%) high mild
  1 (1.00%) high severe

arrow_iter              time:   [44.055 µs 44.165 µs 44.300 µs]
Found 5 outliers among 100 measurements (5.00%)
  5 (5.00%) high mild
```

I'm surprised the Arrow iter is so much faster (and faster than std::iter). My guess is the assertions and alignment checking in our PrimitiveArray make up the difference?

Reply:
Arrow doesn't actually iterator over Option<T> just over T and in this case there's no nulls so arrow is the same as iterating Vec<T>. To get to the level you need monomorphisation of every function call

Reply:
What @robert3005 said, spent a lot of time trying to understand everything they do and while I can't say I have full clarity, Arrow's overall design does make it easier to have fast iterators - both having fully typed arrays and not having to support compression.
I am hopeful that eventually we'll find a more performant way of iterating arrays, but just having a much smaller memory footprint should be useful IMO.